File: nim/search/sentinel_linear_search.nim
import strformat

func sentinelLinearSearch(
  arr: var openArray[int],
  length: int,
  target: int
): int =
  let last = arr[length]
  arr[length] = target

  var index = 0
  while not (arr[index] == target): index += 1

  arr[length] = last

  if index < length or arr[length] == target: 
    return index

  return -1

proc main() =
  let target = 4
  var arr = [3,5,1,8,2,6,4,8,7]
  echo(arr)

  let length = arr.len() - 1
  let res = sentinelLinearSearch(arr, length, target)

  if res == -1:
    echo &"Element {target} not found"
  else:
    echo &"Element {target} found at index {res}"

when isMainModule:
  main()

File: clang/search/sentinel_linear_search.c
#include <stdio.h> 

int sentinel_linear_search(int arr[], int length, int target) { 
	int last_value = arr[length]; 
	arr[length] = target; 

	int index = 0; 
	while (arr[index] != target) { 
		index++; 
	}

	arr[length] = last_value;

	if (index < length || arr[length] == target) { 
		return index;
	} else { 
		return -1;
	} 
} 

int main() { 
	int target = 8; 
	int arr[] = { 3, 5, 1, 8, 2 }; 
	int length = (sizeof(arr) / sizeof(arr[0])) - 1; 
	int result = sentinel_linear_search(arr, length, target); 

	if (result != -1) { 
		printf("Element %d found at index %d\n", target, result); 
	} else { 
		printf("Element %d not found\n", target); 
	} 

	return 0; 
}
